lib/seahorn/CexExeGenerator.cc

```diff
@@ -0,0 +1,575 @@
+
+  // Hook to allocate a dsa node
+  Function *m_memAlloc =
+      cast<Function>(m_harness
+                         ->getOrInsertFunction("__seahorn_mem_alloc",
+                                               Type::getVoidTy(m_context),
+                                               i8PtrTy, i8PtrTy, intTy, intTy)
+                         .getCallee());
+  // Hook to initialize a dsa node
+  Function *m_memInit =
+      cast<Function>(m_harness
+                         ->getOrInsertFunction("__seahorn_mem_init",
+                                               Type::getVoidTy(m_context),
+                                               i8PtrTy, intTy, intTy)
+                         .getCallee());
+
+  for (auto &kv : m_dsa_alloc_map) {
+    unsigned id = kv.first;
+    std::pair<Expr, Expr> limits = kv.second;
+    LOG("cex_verbose", errs() << "Dsa node id=" << id << "\n"
+                              << "start=" << *(limits.first) << " "
+                              << "end=" << *(limits.second) << "\n";);
+
+    MemKv contentVals;
+    Expr defVal;
+
+    // check if we have contents
+    auto it = m_dsa_content_map.find(id);
+    if (it != m_dsa_content_map.end()) {
+      defVal = it->second.first;
+      contentVals = it->second.second;
+      LOG(
+          "cex_verbose", errs() << "default value=" << *(defVal) << "\n";
+          for (auto &kv
+               : contentVals) {
+            errs() << *(kv.first) << "->" << *(kv.second) << "\n";
+          });
+    }
+
+    // __seahorn_mem_alloc(start, end, val, sz);
+    Value *startC = exprToConstant(i8PtrTy, limits.first);
+    Value *endC = exprToConstant(i8PtrTy, limits.second);
+    Value *valC = ConstantInt::get(intTy, 0);
+    if (defVal) {
+      valC = exprToConstant(intTy, defVal);
+    }
+
+    Builder.CreateCall(
+        m_memAlloc, {Builder.CreateBitCast(startC, i8PtrTy),
+                     Builder.CreateBitCast(endC, i8PtrTy), valC,
+                     ConstantInt::get(intTy, m_dl.getTypeStoreSize(intPtrTy))});
+
+    // __seahorn_mem_init(index, val, sz)
+    for (auto &kv : contentVals) {
+      Value *indexC = exprToConstant(i8PtrTy, kv.first);
+      Value *valC = exprToConstant(intTy, kv.second);
+      Builder.CreateCall(
+          m_memInit,
+          {Builder.CreateBitCast(indexC, i8PtrTy), valC,
+           ConstantInt::get(intTy, m_dl.getTypeStoreSize(intPtrTy))});
+    }
+  }
+  Builder.CreateRetVoid();
+}
+
+template <class Trace>
+void CexExeGenerator<Trace>::dumpCexModuleToFile(llvm::StringRef CexFile) {
+  std::error_code error_code;
+  llvm::ToolOutputFile out(CexFile, error_code, sys::fs::F_None);
+  assert(!error_code);
+  verifyModule(*m_harness, &errs());
+  if (CexFile.endswith(".ll"))
+    out.os() << *m_harness;
+  else
+    WriteBitcodeToFile(*m_harness, out.os());
+  out.os().close();
+  out.keep();
+}
+
+template <class Trace>
+Constant *CexExeGenerator<Trace>::exprToConstant(Type *ty, Expr e) {
+  if (isOpX<TRUE>(e)) {
```

Comment:
currently looks like this function is "inferring" type of `LLVM::Constant` to convert to based on `e`, instead of enforcing the input `Type` directly; perhaps we should always try to create a `Constant` of type `ty` here, if somehow input e is not compatible, create a null value and print a debug message.

Reply:
Depends on the client (but this is now under your control). This does sound most reasonable that the output is of a type requested.

There is still a disconnect because types might have different representation in memory. In particular, the memory size of `i1` is `i8`. Probably the client must be aware of that and request appropriate type, or type cast as needed.

Reply:
this suggestion is mostly to resolve original author's concern; the idea is to leave the size of type under client control, so that when it calls `ConstantArray::get(AT, LLVMarray)`, element type of `AT` will not mismatch with `ty`; 
if the client happened to send mismatched Type `ty` and expression `e`, I hope new implementation will make it easier to debug

Reply:
great